pyccel/ast/datatypes.py

```diff
@@ -138,215 +224,449 @@ def __reduce__(self):
         """
         return (self.__class__, ())
 
-class NativeBool(DataType, metaclass=Singleton):
+    def switch_basic_type(self, new_type):
+        """
+        Change the basic type to the new type.
+
+        Change the basic type to the new type. In the case of a FixedSizeType the
+        switch will replace the type completely, directly returning the new type.
+
+        Parameters
+        ----------
+        new_type : PyccelType
+            The new basic type.
+
+        Returns
+        -------
+        PyccelType
+            The new type.
+        """
+        assert isinstance(new_type, FixedSizeType)
+        return new_type
```

Comment:
What is the purpose of this method? It does not use `self` in any way, and it simply adds a type check on `new_type`.

Reply:
The purpose of this method is described here:
https://github.com/pyccel/pyccel/pull/1756/files#diff-7d960d9d73e44e689f3f0d4180a7e7280ac31ef2205bd1e8fc858756f67835b9R73

Its utility is not obvious here as this is the implementation for `FixedSizeType`. This particular implementation is technically not necessary however having it avoids us needing lots of if blocks in the code. E.g.:
```python
class NumpyAbs:
    def __init__(self, arg):
        if arg.rank == 0:
            self._class_type = PythonNativeFloat()
        else:
            self._class_type = arg.class_type.switch_basic_type(PythonNativeFloat())
```

Reply:
That part of the documentation describes the class `HomogeneousContainerType`. This is `FixedSizeType` instead.

Reply:
I do not understand if your example code is the "bad code" which contains the `if` statement that you want to avoid, or if is it the "good code" which avoids some other `if` statement.

Reply:
I have added further explanation to the docs : 790a1a3e8

Reply:
I see. That's neat!